src/attributes.rs

```diff
@@ -0,0 +1,282 @@
+use std::rc::Rc;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Attribute<'a> {
+    pub is_inner: bool,
+    pub content: Rc<AttributeMetaItem<'a>>,
+}
+
+impl<'a> Attribute<'a> {
+    pub fn as_string(&self) -> String {
+        format!(
+            "#{}[{}]",
+            if self.is_inner { "!" } else { "" },
+            self.content.raw_item
+        )
+    }
+
+    pub fn new(raw: &'a str) -> Self {
+        let raw_trimmed = raw.trim();
+        let raw_without_closing = raw_trimmed.strip_suffix(']').unwrap_or_else(|| {
+            panic!(
+                concat!(
+                    "String `{}` cannot be parsed as an attribute ",
+                    "because it is not closed with a square bracket."
+                ),
+                raw_trimmed
+            )
```

Comment:
Here's a slightly more idiomatic way to do what I believe is the same thing:
```suggestion
            panic!(
                "\
String `{raw_trimmed}` cannot be parsed as an attribute \
because it is not closed with a square bracket."
            )
```

Newlines prefixed with `\` are ignored. Also, either the current or the upcoming version of clippy includes a lint that wants all formatting done inline when possible (like `{raw_trimmed}` instead of `"{}", raw_trimmed` above). We'll need to migrate all uses in the codebase to that style when that lint becomes active by default, so let's try to start following it now so don't have to do as much migrating work later.

Reply:
I used `concat!` mainly because strings without indentation seemed a bit ugly to me. But if this is the more idiomatic way, then there is no reason not to follow it.